### ISSUE TITLE
Lose import selection after binding imported transaction to existing transaction

### ISSUE THREAD

**Issue by brownnrl** [hsoft/moneyguru#503]
Often times, I’ll choose to ignore a credit card payment on import as it’s reflected in my bank statement and the date in the credit card statement is offset and I prefer my bank account reflect dates on withdraw.  I have a few other scheduled transactions that I always bind, and I noticed that I would always have to deselect the auto payment after binding.  Otherwise, it ends up re-selected after each bind.  This looks suspiciously like #289 so maybe I broke something when I worked on the import plugin.  I’ll update my version of moneyguru and see if I can re-create with the latest stable release version.

**Comment by brownnrl** [comment]
Moneyguru 2.13.1

1. Open without_some_transactions.moneyguru (which just has a few transactions removed from the attached base.moneyguru).

2. Start an import of sample_import.csv.

3. Bind txn 1 to txn 1, txn 2 to txn 2.

4. Deselect txn 3.

5. Bind txn 4 to txn 4.
==> Here, txn 3 is re-selected which is a bug.

6. De-select txn 3.

7. Break the binding of txn 4, txn 3 is re-selected.

So, the selections aren't maintained after binding.  I run into this when I do an import of my credit card transactions running top to bottom, binding as I go down, and I usually deselect the small overlap and automatic payments since that reflects on my bank statement.

Again, I think I might have broken this one.  Is everything getting a C re-write including imports?  Would you want me to try to fix this aspect, and do a PR in the existing python code?  At the very least, I can try to do a few unit tests illustrating the problem in a PR.

It's not a huge issue, just an annoyance.

[503.tar.gz](https://github.com/hsoft/moneyguru/files/2762233/503.tar.gz)

**Comment by hsoft** [comment]
Thanks for the detailed report, but it seems you're too late, it's already fixed :)

I can reproduce the problem on 2.13.1, but not on master. It's funny because as I worked on the import window a few days ago, I remembered having thought "hum, strange handling of `will_import`" but didn't think of it further as my head was already full of the rather complicated re-factoring I was in.

One thing that would be great, though, if you want, is to write a test for it. tests were passing and the bug was there, so it means that some coverage is missing. Normally, the test should fail on 2.13.1 and pass on master.

**Comment by brownnrl** [comment]
I was getting setup to run unit tests.  I haven't worked with tox before, but I did set up a separate virtual env for python 3.5 and pip installed sip, PyQt5, and py.test.  I have python 3.6 installed, I was just running MG with 3.5 previously so that's where I started.  It complained about a missing pytest-monkeyplus package you had created, so I installed that as well to the virtual env.  When I try to run py.test core, I get the following error:

```
Traceback (most recent call last):
  File "/home/nelson/virtualenvs/mguru/bin/py.test", line 11, in <module>
    sys.exit(main())
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/_pytest/config/__init__.py", line 61, in main
    config = _prepareconfig(args, plugins)
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/_pytest/config/__init__.py", line 196, in _prepareconfig
    pluginmanager=pluginmanager, args=args
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/pluggy/hooks.py", line 284, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/pluggy/manager.py", line 68, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/pluggy/manager.py", line 62, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/pluggy/callers.py", line 203, in _multicall
    gen.send(outcome)
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/_pytest/helpconfig.py", line 93, in pytest_cmdline_parse
    config = outcome.get_result()
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/_pytest/config/__init__.py", line 652, in pytest_cmdline_parse
    self.parse(args)
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/_pytest/config/__init__.py", line 838, in parse
    self._preparse(args, addopts=addopts)
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/_pytest/config/__init__.py", line 784, in _preparse
    self.pluginmanager.load_setuptools_entrypoints("pytest11")
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/pluggy/manager.py", line 268, in load_setuptools_entrypoints
    plugin = ep.load()
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2346, in load
    return self.resolve()
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2352, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 634, in _load_backward_compatible
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/_pytest/assertion/rewrite.py", line 308, in load_module
    six.exec_(co, mod.__dict__)
  File "/home/nelson/virtualenvs/mguru/lib/python3.5/site-packages/pytest_monkeyplus.py", line 12, in <module>
    class monkeyplus(monkeypatch):
TypeError: function() argument 1 must be code, not str
```

That looks like some kind of meta-class issue with the inheritance.  Maybe the version on pip is stale with respect to the current version of monkeypatch in pytest?

I'll try to read through the tox documentation.  I see it auto-scans for existing python interpreters which is neat.  I'll see if I can get it to work for virtual environments.

**Comment by hsoft** [comment]
`pytest-monkeyplus` isn't needed any more. It's been "vendorized" directly in moneyGuru. I'm guessing that its presence conflicts with the vendorized version.

The only requirement now to run tests is `pytest`.

**Comment by brownnrl** [comment]
hmm, I see that a lot of that code is replicated in ```conftest.py```.  I got the monkeyplus name from this error, which occurs after I uinstalled the ```pytest-monkeyplus``` pacage.

```
INTERNALERROR> pluggy.manager.PluginValidationError: unknown hook 'pytest_funcarg__monkeyplus' in plugin <module 'core.tests.conftest' from '/home/nelson/projects/moneyguru/core/tests/conftest.py'>
```

**Comment by hsoft** [comment]
Yes, that's a recent surprise that folks at pytest gave us. They deprecated `pytest_funcarg__` in pytest 4.1. If you run on `master`, you won't have this problem, but if you try to run tests on 2.13.1, install `pytest<4.1`

**Comment by brownnrl** [comment]
Copy that.

**Comment by brownnrl** [comment]
Ok, got it working, thank you!  But that ate all my time for writing a new test for tonight.  :)

I'll put some more time in tomorrow.

**Comment by brownnrl** [comment]
PR #506 ready for review.

**Comment by hsoft** [comment]
Fixed in #506
